### PR TITLE
Create max-area-of-island.java

### PR DESCRIPTION
**Pull request by soyeon51696**
recursion version
I don't know if I can implement the non-recursion version by midnight, so I solved it with the recursive version for now.

### CODE REVIEW

File: study1/week4/max-area-of-island.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+                    area = Math.max(area, calArea(grid, x, y));
+                }
+            }
+        }
+        
+        return area;
+    }
+
+    private int calArea(int[][] grid, int x, int y) {
+        if (!isInRange(grid.length, grid[0].length, x, y) || grid[y][x] == 0) {
```
Comment (h-joo) [inline study1/week4/max-area-of-island.java:25]: Same comment from the floodFill implementation, there's unnecessary comparison here, but I also don't think you can remove this without making this a non-recursive implementation, so it's fine for now.

**Review by h-joo — APPROVED** [review]
Thank you

**Comment by soyeon51696** [comment]
```
class Solution {
    public int maxAreaOfIsland(int[][] grid) {
        //time complexity is O(n*m)
        //space complexity is O(n*m), because the queue can store at most O(n * m) elements
        int m = grid.length; //row 행
        int n = grid[0].length; //column 열
        int maxArea = 0;
        int[][] directions = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};

        for (int x = 0; x < m; x++) { //x는 행 개수까지
            for (int y = 0; y < n; y++) { //y는 열 개수까지
                if (grid[x][y] == 1) {
                    int area = 0;
                    
                    Queue<int[]> queue = new LinkedList<>();
                    queue.offer(new int[]{x, y});

                    while (!queue.isEmpty()) {
                        int[] point = queue.poll();
                        int curX = point[0];
                        int curY = point[1];

                        grid[curX][curY] = 0;
                        area++;

                        if (curY > 0  && grid[curX][curY - 1] == 1) {
                            queue.add(new int[]{curX, curY - 1});
                        }
                        if (curY < n - 1 && grid[curX][curY + 1] == 1) {
                            queue.add(new int[]{curX, curY + 1});
                        }
                        if (curX > 0  && grid[curX - 1][curY] == 1) {
                            queue.add(new int[]{curX - 1, curY});
                        }
                        if (curX < m - 1 && grid[curX + 1][curY] == 1) {
                            queue.add(new int[]{curX + 1, curY});
                        }  

                    }

                maxArea = Math.max(maxArea, area);
                }
            }
        } 
        return maxArea;
    }
}
```

I've modified it up to this point, but some tests don't pass...

**Comment by soyeon51696** [comment]
```
class Solution {
    public int maxAreaOfIsland(int[][] grid) {
        //time complexity is O(n*m)
        //space complexity is O(n*m), because the queue can store at most O(n * m) elements
        int m = grid.length; //row 행
        int n = grid[0].length; //column 열
        int maxArea = 0;

        for (int x = 0; x < m; x++) { //x는 행 개수까지
            for (int y = 0; y < n; y++) { //y는 열 개수까지
                if (grid[x][y] == 1) {
                    int area = 0;
                    
                    Queue<int[]> queue = new LinkedList<>();
                    queue.offer(new int[]{x, y});
                    grid[x][y] = 0;

                    while (!queue.isEmpty()) {
                        int[] point = queue.poll();
                        int curX = point[0];
                        int curY = point[1];
                        area++;

                        if (curY > 0  && grid[curX][curY - 1] == 1) {
                            queue.add(new int[]{curX, curY - 1});
                            grid[curX][curY - 1] = 0; 
                        }
                        if (curY < n - 1 && grid[curX][curY + 1] == 1) {
                            queue.add(new int[]{curX, curY + 1});
                            grid[curX][curY + 1] = 0; 
                        }
                        if (curX > 0  && grid[curX - 1][curY] == 1) {
                            queue.add(new int[]{curX - 1, curY});
                            grid[curX - 1][curY] = 0; 
                        }
                        if (curX < m - 1 && grid[curX + 1][curY] == 1) {
                            queue.add(new int[]{curX + 1, curY});
                            grid[curX + 1][curY] = 0;
                        }

                    }

                maxArea = Math.max(maxArea, area);
                }
            }
        } 
        return maxArea;
    }
}
```

solved!

**Pull request merged**